## Basics of logistic regression

### Import the relevant libraries

In [1]:
import numpy as np
import pandas as pd
import statsmodels.api as sm
import matplotlib.pyplot as plt
import seaborn as sns
sns.set()

### Data

In [2]:
raw_data = pd.read_csv('C://Users//irem//Desktop//The Data Science Course 2021 - All Resources//Part_5_Advanced_Statistical_Methods_(Machine_Learning)//S36_L235//2.01. Admittance.csv')
raw_data

SAT Admitted
0    1363       No
1    1792      Yes
2    1954      Yes
3    1653       No
4    1593       No
..    ...      ...
163  1722      Yes
164  1750      Yes
165  1555       No
166  1524       No
167  1461       No

[168 rows x 2 columns]

In [3]:
data = raw_data.copy()
data['Admitted'] = data['Admitted'].map({'Yes': 1, 'No': 0})
data

SAT  Admitted
0    1363         0
1    1792         1
2    1954         1
3    1653         0
4    1593         0
..    ...       ...
163  1722         1
164  1750         1
165  1555         0
166  1524         0
167  1461         0

[168 rows x 2 columns]

In [5]:
y = data['Admitted']
x1 = data['SAT']

### Regression

In [6]:
x = sm.add_constant(x1)
reg_log = sm.Logit(y, x)
results_log = reg_log.fit()

Optimization terminated successfully.
         Current function value: 0.137766
         Iterations 10


In [7]:
results_log.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:               Admitted   No. Observations:                  168
Model:                          Logit   Df Residuals:                      166
Method:                           MLE   Df Model:                            1
Date:                Tue, 04 Jan 2022   Pseudo R-squ.:                  0.7992
Time:                        10:55:39   Log-Likelihood:                -23.145
converged:                       True   LL-Null:                       -115.26
Covariance Type:            nonrobust   LLR p-value:                 5.805e-42
==============================================================================
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
const        -69.9128     15.737     -4.443      0.000    -100.756     -39.070
SAT            0.0420      0.009      4.454      0.000       0.024       0.060
==============================================================================

Possibly complete quasi-separation: A fraction 0.27 of observations can be
perfectly predicted. This might indicate that there is complete
quasi-separation. In this case some parameters will not be identified.
"""

### MLE (Maximum likelihood estimation)
Likelihood function is a function which estimates how likely it is that the model at hand describes the real underlying relationship of the variables.

The bigger the likelihood function, the higher the probability that our model is correct.

MLE tries to maximize the likelihood function.
The computer is going through different values, until it finds a model, for which the likelihood is the highest.
When it can no longer improve it, it will just stop the optimization.

#### Log-likelihood:
More popular metric of conducting MLE.
Log-likelihood value is almost but not always negative, and the bigger it is, the better.

#### LL-null (log-likelihood null):
The log-likelihood of a model which has no independent variables.
We can compare our model with no explanatory variables since we may see whether the model has any explanatory power.
Similar to F-statistic in linear regression.

#### LLR p-value:
Log-likelihood ratio test, it measures if our model is statistically different from LL-null (a useless model).

#### Pseudo R-Squared:
This is also called "McFadden's R-Squared", and according to McFadden, a good pseudo R-squared is somewhere between 0.2 and 0.4.
This measure is mostly useful for comparing variations of the same model. Different models will have completely different and incomparable Pseudo R-squares. 

The model becomes as following:

$log(\frac{\pi}{1-\pi}) = -69.91 + 0.042*SAT$

where $\pi$ is the probability of the event occurring, and $1-\pi$ is the probabilitty of event not occurring.

So; $log(odds) = -69.91 + 0.042*SAT$
 